compiler/ETA/TypeCheck/TcEnv.hs

```diff
@@ -784,12 +785,26 @@ newDFunName :: Class -> [Type] -> SrcSpan -> TcM Name
 newDFunName clas tys loc
   = do  { is_boot <- tcIsHsBootOrSig
         ; mod     <- getModule
-        ; let info_string = occNameString (getOccName clas) ++
-                            concatMap (occNameString.getDFunTyKey) tys
-        ; dfun_occ <- chooseUniqueOccTc (mkDFunOcc info_string is_boot)
+        ; dfun_occ <- chooseUniqueOccTc $ mkDFunOcc (mkInfoString clas tys) is_boot
         ; newGlobalBinder mod dfun_occ loc }
 
 {-
+It may be advantageous at some point to build a string for a type which actually
+encompasses the structure of the type like parethesizing.
+Even in the new mkInfoString something like `a (b c)` and `(a b) c` will result
```

Comment:
You can handle parenthesis with Z-encoding, but I'm not sure it's worth it at this point. We can do that if it seriously becomes an issue.